modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -415,7 +415,17 @@ public CompletableFuture<T> apply(T r, Throwable e) {
                         throw (RuntimeException) e;
                     }); // Preserve failed state.
                 } else {
-                    tx0.enlistResultFuture(fut);
+                    try {
+                        tx0.enlistResultFuture(fut);
+                    } catch (TransactionException enlistResultException) {
+                        return tx0.rollbackAsync().handle((ignored, err) -> {
```

Comment:
Why are you rolling back the transaction if it is already finished?

Reply:
Because it's required to cleanup locks and writeIntents for operation that failed to enlistResultFuture.

Reply:
So that will not happen because the transaction is already finished.

Reply:
You are right we need an escape path in order to rollback partially finished transaction.